automata/fa/dfa.py

```diff
@@ -229,7 +229,7 @@ def _check_for_input_rejection(self, current_state):
                 'the DFA stopped on a non-final state ({})'.format(
                     current_state))
 
-    def read_input_stepwise(self, input_str):
+    def read_input_stepwise(self, input_str, check=True):
```

Comment:
Oh yeah I forgot about this actually. I feel like this function could just read and not care about the acceptance state honestly. If not, then we should have a function which does that.

Reply:
I think it's up to @caleb531, but ultimately I agree that it's definitely useful to have a function that gives the states as the input is being read without throwing an exception.

Reply:
I hadn't considered partially defined DFAs. Everything I have implemented for DFAs has assumed they are well and fully defined. Like I said, I had forgotten I made this change, it was simply to make use of the function which reads stepwise. 

Regarding behaviour I would think that DFAs constructed with allow_partial have a specific error state which is either implicit or added explicitly in the constructor. The use case is exactly what I'm doing here, I'm gathering the states visited from reading the input step by step.` In this use case I don't care if I end up at a final state or not since I intend to explore further. If the user wants to check for acceptance then that can be done by using another function. If the user wants to read input stepwise then this function can be used. If the user wants to do both then the user can start by calling this function then just check if the state is in the final states.

So ideally I would like this function to default to the behavior I use it for here, because then it is doing what the name says. Checking for acceptance is not implied by the name I feel.

Reply:
@Tagl I understand, but my preferred approach would be to have the user wrap their `read_input_stepwise` call in a try..except block that catches `RejectionException`. That is why this library exposes so many specialized exceptions—to allow these methods to be easily try..except'able.

Granted, because the exception is raised before the function terminates, this prevents you from converting the result of `read_input_stepwise` to a list (e.g. in a list comp) if the input is rejected. The only way around this it to store each value in a list as they are yielded (because the yielding still happens before the rejection check in the source code).

```python
try:
    states = []
    for state in dfa.read_input_stepwise('0011'):
        states.append(state)
except RejectionException:
    pass
print(states)
```

^ I would agree—not my favorite workaround.

If we implement this flag, @eliotwrobson Is right that the flag should be implemented by all Automaton subtypes, since I have explicitly declared `read_input_stepwise` as an abstract method in the top-level `Automaton` class. And I would still prefer the name `allow_rejection` or maybe even better `ignore_rejection` (if you pick the latter, just make sure to invert the logic). If we go this route of some sort of flag, I think I would prefer the current behavior to be the default, in order to avoid making _too_ many backwards-incompatible changes. 😅

Reply:
I just generally disagree with exceptions being used unless in exceptional cases, and this does not feel that way. However, changing that would require a rewrite of many things in the library and changes the public API so I think it's easier to just define a helper which `DFA.read_input_stepwise` uses which does just the reading part, without checking acceptance.

Reply:
I'm also fine with having the extra parameter with one of your proposed names, I think any type of automaton can just yield a sentinel error state (`None`?) if it is an implicit one for the undefined transitions.

Reply:
@Tagl I suppose, although for a partial DFA, as soon as you hit a transition with one of these sentinel "end states" (like `None`), you won't be able to advance the DFA past that point, because the sentinel value is not a real state. So in addition to updating `_get_next_current_state` / `_get_next_current_states` for all Automaton types to properly produce `None`, we will likely need to add add'l logic handle those `None` values.

Reply:
I'll have to try to use the partial construction to see what happens.

Reply:
Just updated the implementation and left the default rejection behavior in so there's no breaking API change. Take a look.